[Distributed] DeleteKey API for c10d TCP Store

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45223 [Distributed] Adding Python tests for the TCPStore getNumKeys and deleteKey
* **#43963 [Distributed] DeleteKey API for c10d TCP Store**
* #43962 [Distributed] getNumKeys API to c10d TCPStore

This PR adds a deleteKey API to the TCP Store, which deletes the specified key from the store. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

osalpekar committed:
[Distributed] DeleteKey API for c10d TCP Store

Added a DeleteKey API for the TCP Store

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


Added a DeleteKey API for the TCP Store

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


Added a DeleteKey API for the TCP Store

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


Added a DeleteKey API for the TCP Store

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


Added a DeleteKey API for the TCP Store

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

pritamdamania87 commented on caffe2/distributed/file_store_handler.cc:
```diff
@@ -127,6 +127,10 @@ int64_t FileStoreHandler::getNumKeys() {
 }
 
+void FileStoreHandler::deleteKey(const std::string& /* unused */) {
+  CHECK(false) << "deleteKey not implemented for FileStoreHandler";
```
`TORCH_CHECK`?

pritamdamania87 commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -40,6 +40,10 @@ void testHelper(const std::string& prefix = "") {
     EXPECT_EQ(numKeys, 5);
+
+    serverStore->deleteKey("key0");
+    numKeys = serverStore->getNumKeys();
```
nit: probably also test `getKey` returns an error.

pritamdamania87 replied:
Can we add a test here for calling `getKey()` on a delete key?

pritamdamania87 commented on torch/lib/c10d/Store.hpp:
```diff
@@ -30,6 +30,8 @@ class Store {
 
   virtual int64_t add(const std::string& key, int64_t value) = 0;
 
+  virtual void deleteKey(const std::string& key) = 0;
```
Usually a delete API returns true/false about whether the key was removed or not. I think this should return a bool.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -303,6 +307,10 @@ They are used in specifying strategies for reduction collectives, e.g.,
               "add",
```
All of these would need docs for python users. We can do that in a separate PR.

pritamdamania87 reviewed (changes requested):
Requesting changes since I think the API should return a bool

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


This PR adds a deleteKey API to the TCP Store, which deletes the specified key from the store. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


This PR adds a deleteKey API to the TCP Store, which deletes the specified key from the store. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

osalpekar committed:
Update on "[Distributed] DeleteKey API for c10d TCP Store"


This PR adds a deleteKey API to the TCP Store, which deletes the specified key from the store. This API will be useful for some applications related to debug logging in ProcessGroupNCCL going forward. This PR also adds some C++ tests for this API and Python tests are added in #45223. We will build on this functionality in the future by implementing this API for FileStore, HashStore, RedisStore, and ZeusStore.

Differential Revision: [D23009117](https://our.internmc.facebook.com/intern/diff/D23009117/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D23009117/)!

[ghstack-poisoned]

ngimel commented:
Reverting, as it is breaking OSS build
```
08:04:53 /var/lib/jenkins/workspace/caffe2/distributed/redis_store_handler.cc: In member function ‘virtual bool caffe2::RedisStoreHandler::deleteKey(const string&)’:
08:04:53 /var/lib/jenkins/workspace/caffe2/distributed/redis_store_handler.cc:86:1: error: no return statement in function returning non-void [-Werror=return-type]
08:04:53  }
08:04:53  ^
08:04:53 cc1plus: some warnings being treated as errors
08:04:53 caffe2/CMakeFiles/torch_cpu.dir/build.make:10370: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/distributed/redis_store_handler.cc.o' failed
08:04:53 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/distributed/redis_store_handler.cc.o] Error 1
08:04:53 make[2]: *** Waiting for unfinished jobs....
08:04:58 CMakeFiles/Makefile2:6884: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed
08:04:58 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2
08:04:58 Makefile:140: recipe for target 'all' failed
08:04:58 make: *** [all] Error 2
```
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/pytorch-linux-bionic-rocm3.7-py3.6-build/5732//console
